third_party/nvfuser/csrc/expr_simplifier.cpp

```diff
@@ -1188,12 +1185,29 @@ bool isMultipleOf(Val* x, Val* y) {
   return sym_algebra::divideFactorized(lhs, rhs) != nullptr;
 }
 
-bool hasCompatibleSign(Val* x, Val* y, const VarInfoMap& var_info) {
-  return isNonNegative(x, var_info) && isNonNegative(y, var_info);
+bool hasCompatibleSign(Val* x, Val* y, const Context& context) {
+  return isNonNegative(x, context) && isNonNegative(y, context);
 }
 
 } // namespace prove
 
+namespace {
+
+// If we can not prove `value` to be zero, then treat it as non-zero, unless we
+// want to preserve error. This is useful for simplifications like (a * b) / b.
+// If we want to preserve error, then we can simplify the above expr to `a` iff
+// we can prove that b is not zero. If we don't care about error case, then
+// having `b` in the denominator already indicates that `b` is non-zero. So we
+// can just do the simplification without worrying about `b` being zero.
+bool notSurelyZero(Val* value, const Context& context) {
+  if (context.preserveError()) {
+    return prove::isNonZero(value, context);
+  }
+  return !foldConstants(value)->isZero();
```

Comment:
Does folding of `(a * b) / b` to `b` happen in `foldConstants(value)`? Does `foldConstants` fold the expression no matter if `b` is zero or not?

Reply:
No, `foldConstants` is just [constant folding](https://en.wikipedia.org/wiki/Constant_folding), that is, changing `3 - 3` into `0`.